pkg/cmd/extension/manager.go

```diff
@@ -320,41 +381,82 @@ func (m *Manager) Upgrade(name string, force bool, stdout, stderr io.Writer) err
 	someUpgraded := false
 	for _, f := range exts {
 		if name == "" {
-			fmt.Fprintf(stdout, "[%s]: ", f.Name())
+			fmt.Fprintf(m.io.Out, "[%s]: ", f.Name())
 		} else if f.Name() != name {
 			continue
 		}
 
 		if f.IsLocal() {
 			if name == "" {
-				fmt.Fprintf(stdout, "%s\n", localExtensionUpgradeError)
+				fmt.Fprintf(m.io.Out, "%s\n", localExtensionUpgradeError)
 			} else {
 				err = localExtensionUpgradeError
 			}
 			continue
 		}
 
-		var cmds []*exec.Cmd
-		dir := filepath.Dir(f.Path())
-		if force {
-			fetchCmd := m.newCommand(exe, "-C", dir, "--git-dir="+filepath.Join(dir, ".git"), "fetch", "origin", "HEAD")
-			resetCmd := m.newCommand(exe, "-C", dir, "--git-dir="+filepath.Join(dir, ".git"), "reset", "--hard", "origin/HEAD")
-			cmds = []*exec.Cmd{fetchCmd, resetCmd}
-		} else {
-			pullCmd := m.newCommand(exe, "-C", dir, "--git-dir="+filepath.Join(dir, ".git"), "pull", "--ff-only")
-			cmds = []*exec.Cmd{pullCmd}
+		binManifestPath := filepath.Join(filepath.Dir(f.Path()), manifestName)
+		if _, e := os.Stat(binManifestPath); e == nil {
+			err = m.upgradeBin(f)
```

Comment:
I don't know if it needs to be addressed now, since this behavior already existed, but it would be nice to not overwrite errors that may have occurred upgrading previous extensions. That way the user will know all the extensions that failed to upgrade.

Reply:
yeah, i was confused by that code and would love it to be cleaned up. given that i'm out starting tomorrow i might punt though.